src/Account/Service/UserFollowService.ts

```diff
@@ -0,0 +1,334 @@
+        );
+      }
+    }
+
+    // TODO 优化为先写缓存再同步DB
+
+    const createdAt = new Date();
+    await this.userFollowModel.create({
+      userId,
+      targetId,
+      createdAt,
+    });
+
+    await this.redisService
+      .getClient()
+      .zadd(RedisKeys.UserFollow(userId), createdAt.getTime(), targetId);
+
+    const targetUserFansCount = await this.redisService
+      .getClient()
+      .zcard(RedisKeys.UserFans(targetId));
+
+    if (
+      targetUserFansCount >=
+      this.configService.get<number>('app.relationship.maxFansCacheCount')
+    ) {
+      const last = await this.redisService
+        .getClient()
+        .zrangebyscore(
+          RedisKeys.UserFans(userId),
+          '-inf',
+          '+inf',
+          'LIMIT',
+          0,
+          1,
+        );
+
+      // remove last
+      await this.redisService
+        .getClient()
+        .zrem(RedisKeys.UserFans(targetId), last);
+    }
+
+    await this.redisService
+      .getClient()
+      .zadd(RedisKeys.UserFans(targetId), createdAt.getTime(), userId);
+
+    this.logger.info(`add ${userId} to ${targetId} fans list`);
+    await this.redisService
+      .getClient()
+      .hincrby(RedisKeys.UserCache(userId), 'followCount', 1);
+
+    await this.redisService
+      .getClient()
+      .hincrby(RedisKeys.UserCache(targetId), 'fansCount', 1);
+  }
+
+  async unFollowAll(userId: string) {
+    const users = await this.userFollowModel.find(
+      {
+        userId: userId,
+      },
+      { targetId: 1 },
+    );
+
+    if (users.length === 0) {
+      return;
+    }
+
+    await this.userFollowModel.deleteMany({ userId });
+
+    // to mq task ↓
+    // delete cache
+
+    const pipeline = this.redisService.getClient().pipeline();
+    pipeline.hincrby(RedisKeys.UserCache(userId), 'followCount', -users.length);
+
+    for (const target of users) {
+      pipeline.zrem(RedisKeys.UserFans(target.targetId), userId);
+      pipeline.hincrby(RedisKeys.UserCache(target.targetId), 'fansCount', -1);
+    }
+
+    const res = await pipeline.exec();
+    this.logger.info(`deleted all follow ${userId}`);
+  }
+
+  async unFollow(userId: string, targetId: string) {
+    await this.userFollowModel.deleteOne({
+      userId,
+      targetId,
+    });
+
+    // delete cache
+
+    // follow 有 maxFollow 限制, 删一个加一个理论上不存在一致性问题
+    await this.redisService
+      .getClient()
+      .zrem(RedisKeys.UserFollow(userId), targetId);
+
+    // 删除之后会导致空缺, 先检查数量, 不够就补一个
+    // 如果恶性事件, 一堆人取关
+    // 强业务相关, // todo
+
+    // 先看看在不在缓存里面
+    const remRes = await this.redisService
+      .getClient()
+      .zrem(RedisKeys.UserFans(targetId), userId);
+
+    if (remRes !== 0) {
+      const targetUserFansCount = await this.redisService
+        .getClient()
+        .zcard(RedisKeys.UserFans(targetId));
+      if (
+        targetUserFansCount <
+        this.configService.get<number>('app.relationship.maxFansCacheCount')
+      ) {
+        await this.cacheFans(userId);
```

Comment:
why it is necessary to refresh the fans list?

Reply:
In order to avoid inconsistent caches, of course, you can also add the corresponding number of caches. My recommendation is to use mq to refresh asynchronously